Dino

lframosferreira opened the pull request:
Jogo do dinossauro do chrome implementado. De forma geral, o jogo funciona bem, mas ainda não imlpementa coisas importantes como animações dos dinossauros ao se movimentar (bater de asas do ptera e pézinhos do Trex).

A sprite do T rex agachado tá com um sombreamento estranho em vola dela que não consegui identificar a origem, mas não é anda que prejudique jogabilidade, só visual.

Um ponto a se atentar é a lógica de colisão desse jogo, que também interfere nos outros. Em jogos como o Snake a colisao entre Sprites "perfeitas" (quadrados, retângulos, etc) é bem trivial. Aqui, a colisão deveria ser feita nos pixels em que as entidades existem de fato, não nesse quadrado da Sprite. No entanto, isso não ocorre, gerando situações estranhas em que aparentemente não há colisão, mas há sim um overlap de sprites. 

Não sei se o Bevy já tem uma implementação pra lidar com isso. No PyGame, tinha um conceito de máscaras pra tratar esse tipo de problema. Segue o link:
https://www.pygame.org/docs/ref/mask.html

lframosferreira committed:
adicionei nuvens

lframosferreira committed:
adicionei aumento de velocidade conforme o tempo passa

lframosferreira committed:
fix: reset dos componentes quando rola game over

lframosferreira committed:
tentando corrigir o movimento de agachamento do dino

lframosferreira committed:
tentando adicionar sprite

lframosferreira committed:
deixando mais bonitin

lframosferreira committed:
removendo alguns warnings

lframosferreira committed:
algumas correcoes na movimentação

lframosferreira committed:
decrementei o spawn de obstaculos pra dificultar um cado

igorlfs commented:
Alguém não gosta de instalar hooks de pre-commit

igorlfs committed:
cargo fmt

igorlfs commented on dino/src/game/cloud/mod.rs:
```diff
@@ -0,0 +1,26 @@
+mod systems;
+
+use self::resources::*;
+use self::systems::*;
```
Deixa os módulos em cima, tira o self (infelizmente o cargo fmt não padroniza isso)

igorlfs commented on dino/src/game/cloud/systems.rs:
```diff
@@ -0,0 +1,54 @@
+use bevy::prelude::*;
+use bevy::window::PrimaryWindow;
+use rand::prelude::*;
+
```
remover essa linha (e formatar) pro `cargo fmt` organizar

igorlfs commented on dino/src/game/dinosaur/resources.rs:
```diff
@@ -16,3 +16,14 @@ impl Default for DinoVerticalMovement {
     }
 }
+
+#[derive(Resource)]
```
Deriva o Default

igorlfs commented on dino/src/game/dinosaur/systems.rs:
```diff
@@ -37,7 +40,7 @@ pub fn handle_jump(
             }
         } else {
-            if keyboard_input.just_pressed(KeyCode::Space) {
+            if keyboard_input.just_pressed(KeyCode::Space) && !dino_down.is_down {
```
Usa else if ao invés de fazer um if dentro do else

igorlfs commented on dino/src/game/dinosaur/systems.rs:
```diff
@@ -48,41 +51,61 @@ pub fn handle_jump(
-                return;
+            } else {
+                score.value += 1;
             }
```
O comportamento atual disso é incrementar a pontuação a cada frame que não existe colisão, para cada obstáculo. É realmente isso que a gente quer fazer? É mais parecido com o jogo original do dino (em comparação ao flappy bird, que a gente só aumenta a pontuação em 1, a cada objeto que fica pra trás), mas acho que tem jeitos melhores de fazer uma pontuação contínua -- por exemplo, ter um recurso que conta o tempo desde que o início do jogo (tratar restart e quit)

igorlfs commented on dino/src/systems.rs:
```diff
@@ -1,38 +1,38 @@
-    }
+pub fn reset_obstacle_speed(mut obstacle_speed: ResMut<ObstacleSpeed>) {
+    obstacle_speed.speed = OBSTACLE_INITIAL_SPEED;
 }
 
-pub fn resume_game(
-    keyboard_input: Res<Input<KeyCode>>,
-    app_state: Res<State<AppState>>,
-    mut next_app_state: ResMut<NextState<AppState>>,
+pub fn despawn_obstacles(
+    mut commands: Commands,
+    mut obstacle_query: Query<Entity, With<Obstacle>>,
 ) {
-    if keyboard_input.just_pressed(KeyCode::Escape) && app_state.get() == &AppState::Pause {
-        next_app_state.set(AppState::InGame);
+    for obstacle_entity in obstacle_query.iter_mut() {
+        commands.entity(obstacle_entity).despawn();
     }
 }
```
Não é melhor essas funções ficarem nos seus respectivos módulos?

igorlfs commented on dino/src/systems.rs:
```diff
@@ -1,38 +1,38 @@
+    if let Ok((mut dinosaur_transform, dinosaur_entity)) = dinosaur_query.get_single_mut() {
+        dinosaur_transform.translation.y = DINO_INITIAL_Y_POS;
+        commands.entity(dinosaur_entity).insert(SpriteBundle {
+            transform: dinosaur_transform.clone(),
```
```suggestion
            transform: *dinosaur_transform,
```

igorlfs reviewed (changes requested):
Você poderia trocar o título da janela pra "Bevy Dinosaur"? A cor de fundo fica meio ruim com a crate common, mas depois mudo isso

igorlfs commented on dino/src/game/score/mod.rs:
```diff
@@ -11,7 +11,6 @@ pub struct ScorePlugin;
         app.init_resource::<Score>()
-            .add_systems(OnEnter(AppState::InGame), insert_score)
-            .add_systems(OnExit(AppState::InGame), remove_score);
+            .add_systems(OnEnter(AppState::InGame), insert_score);
```
Isso tem um bug herdado do snake: ao pausar o jogo, o score é resetado. No snake isso foi corrigido em #12 e a ideia é só inserir no evento de Startup.

lframosferreira committed:
fix pr

lframosferreira committed:
fix pr

lframosferreira committed:
nova logica de incremento do score conforme o tempo passa

igorlfs commented on dino/src/game/cloud/mod.rs:
```diff
@@ -0,0 +1,26 @@
+use common::*;
+
+use resources::*;
+use systems::*;
```
```suggestion
use bevy::prelude::*;
use common::*;
use resources::*;
use systems::*;
```

igorlfs commented on dino/src/game/cloud/systems.rs:
```diff
@@ -0,0 +1,53 @@
+pub const CLOUD_SPEED: f32 = 450.0;
+
+pub fn tick_cloud_spawn_timer(mut cloud_spawn_timer: ResMut<CloudSpawnTimer>, time: Res<Time>) {
+    cloud_spawn_timer.timer.tick(time.delta());
```
A gente poderia introduzir alguma aleatoriedade gerando um número aleatório e verificando se ele é menor que um limite.

lframosferreira replied:
Mas porque?

igorlfs replied:
Só pra introduzir um pouco de aleatoriedade

lframosferreira replied:
Acho que fica um pouco de overkill ksksks

igorlfs commented on dino/src/game/cloud/systems.rs:
```diff
@@ -0,0 +1,53 @@
+            commands.entity(cloud_entity).despawn();
+        }
+    }
+}
```
```suggestion
pub fn despawn_clouds_over_time(
    mut commands: Commands,
    cloud_query: Query<(&Transform, Entity), With<Cloud>>,
) {
    for (cloud_transform, cloud_entity) in cloud_query.iter() {
        if cloud_transform.translation.x < 0.0 {
            commands.entity(cloud_entity).despawn();
        }
    }
}
```

Evite usar variáveis mutáveis (se não for necessário)

igorlfs commented on dino/src/game/dinosaur/mod.rs:
```diff
@@ -1,31 +1,33 @@
-use bevy::prelude::*;
-
-mod components;
+pub mod components;
```
```suggestion
mod components;
```
Agora esse módulo não precisa ser público

igorlfs commented on dino/src/game/dinosaur/systems.rs:
```diff
@@ -36,53 +39,84 @@ pub fn handle_jump(
+    if let Ok((dinosaur_image_handle, dinosaur_transform)) = dinosaur_query.get_single() {
+        for (obstacle_image_handle, obstacle_transform) in obstacle_query.iter() {
+            let dinosaur_dimensions: Vec2 = assets.get(dinosaur_image_handle).unwrap().size();
+            let obstacle_dimensions: Vec2 = assets.get(obstacle_image_handle).unwrap().size();
```
```suggestion
            let dinosaur_dimensions = assets.get(dinosaur_image_handle).unwrap().size();
            let obstacle_dimensions = assets.get(obstacle_image_handle).unwrap().size();
```
Não anote tipos que podem ser inferidos

igorlfs commented on dino/src/game/dinosaur/systems.rs:
```diff
@@ -36,53 +39,84 @@ pub fn handle_jump(
+                    texture: asset_server.load("sprites/dino/dino_down_1.png"),
+                    ..default()
+                });
+                dino_down.is_down = true;
```
Você precisa resetar o `is_down` quando o jogo reiniciar, porque se você morrer agachado, não vai conseguir pular (tente).

igorlfs commented on dino/src/game/dinosaur/systems.rs:
```diff
@@ -36,53 +39,84 @@ pub fn handle_jump(
+    mut dino_down: ResMut<DinoDown>,
 ) {
-    if let Ok(mut dinosaur_transform) = dinosaur_query.get_single_mut() {
+    if let Ok(dinosaur_entity) = dinosaur_query.get_single_mut() {
```
```suggestion
    dinosaur_query: Query<Entity, With<Dinosaur>>,
    keyboard_input: Res<Input<KeyCode>>,
    asset_server: Res<AssetServer>,
    dinosaur_vertical_movement: Res<DinoVerticalMovement>,
    mut dino_down: ResMut<DinoDown>,
) {
    if let Ok(dinosaur_entity) = dinosaur_query.get_single() {
```

igorlfs commented on dino/src/game/dinosaur/systems.rs:
```diff
@@ -36,53 +39,84 @@ pub fn handle_jump(
+    if let Ok((mut dinosaur_transform, dinosaur_entity)) = dinosaur_query.get_single_mut() {
+        dinosaur_transform.translation.y = DINO_INITIAL_Y_POS;
+        commands.entity(dinosaur_entity).insert(SpriteBundle {
+            transform: dinosaur_transform.clone(),
```
```suggestion
            transform: *dinosaur_transform,
```

igorlfs commented on dino/src/game/mod.rs:
```diff
@@ -1,10 +1,12 @@
-mod obstacle;
-mod score;
+pub mod obstacle;
+pub mod score;
```
Nenhum desses módulos precisa ser público

igorlfs commented on dino/src/game/obstacle/mod.rs:
```diff
@@ -1,26 +1,36 @@
 
-use self::resources::ObstacleSpawnTimer;
+use self::resources::{ObstacleSpawnTimer, ObstacleSpeed};
 use self::systems::*;
```
```suggestion
pub mod components;
mod resources;
mod systems;

use bevy::prelude::*;
use common::*;
use resources::{ObstacleSpawnTimer, ObstacleSpeed};
use systems::*;

const OBSTACLE_INITIAL_SPEED: f32 = 600.0;
const OBSTACLE_SPEED_INCREASE_RATE: f32 = 100.0;
```

igorlfs commented on dino/src/game/obstacle/resources.rs:
```diff
@@ -1,6 +1,8 @@
 use bevy::prelude::*;
 
-pub const OBSTACLE_SPAWN_TIME: f32 = 2.0;
+use super::OBSTACLE_INITIAL_SPEED;
+
+pub const OBSTACLE_SPAWN_TIME: f32 = 1.5;
```
```suggestion
use super::OBSTACLE_INITIAL_SPEED;
use bevy::prelude::*;

const OBSTACLE_SPAWN_TIME: f32 = 1.5;
```

igorlfs commented on dino/src/game/obstacle/systems.rs:
```diff
@@ -1,12 +1,13 @@
 use bevy::window::PrimaryWindow;
+use rand::rngs::ThreadRng;
 use rand::Rng;
 
-use super::OBSTACLE_SPEED;
-use crate::game::dinosaur::DINO_INITIAL_Y_POS;
-
 use super::components::{Obstacle, ObstacleKind};
-use super::resources::ObstacleSpawnTimer;
+use super::resources::{ObstacleSpawnTimer, ObstacleSpeed};
+use super::{OBSTACLE_INITIAL_SPEED, OBSTACLE_SPEED_INCREASE_RATE};
+use crate::game::dinosaur::DINO_INITIAL_Y_POS;
+use crate::game::score::resources::Score;
```
```suggestion
use super::components::{Obstacle, ObstacleKind};
use super::resources::{ObstacleSpawnTimer, ObstacleSpeed};
use super::{OBSTACLE_INITIAL_SPEED, OBSTACLE_SPEED_INCREASE_RATE};
use crate::game::dinosaur::DINO_INITIAL_Y_POS;
use crate::game::score::resources::Score;
use bevy::prelude::*;
use bevy::window::PrimaryWindow;
use rand::rngs::ThreadRng;
use rand::Rng;
```

igorlfs commented on dino/src/game/obstacle/systems.rs:
```diff
@@ -25,9 +26,12 @@ pub fn spawn_obstacles_over_time(
         let window: &Window = window_query.get_single().unwrap();
         let obstacle_kind: ObstacleKind = rand::random();
         let amount: u32 = rand::thread_rng().gen_range(1..4);
+        let mut rng: ThreadRng = rand::thread_rng();
```
Remova as anotações de tipo

igorlfs commented on dino/src/game/score/resources.rs:
```diff
@@ -1,6 +1,21 @@
-use bevy::prelude::Resource;
+use bevy::prelude::{Resource, Timer, TimerMode};
```
Acho que se a gente for usar mais de uma coisa do bevy, já compensa fazer o star import

igorlfs commented on dino/src/game/score/resources.rs:
```diff
@@ -1,6 +1,21 @@
+impl Default for ScoreUpdateTimer {
+    fn default() -> Self {
+        Self {
+            timer: Timer::from_seconds(SCORE_UPDATE_TIME, TimerMode::Repeating),
```
Talvez ficaria mais legal contar milissegundos (e dividir por alguma coisa)?

igorlfs commented on dino/src/main.rs:
```diff
@@ -16,13 +15,12 @@ fn main() {
             }),
             ..default()
         }))
         .add_plugins(DebugLinesPlugin::default())
```
```suggestion
```
Não estamos usando as linhas nesse jogo

igorlfs commented on dino/src/main.rs:
```diff
@@ -1,13 +1,12 @@
+use game::GamePlugin;
 
-use systems::*;
+pub const WINDOW_X: f32 = 1000.0;
+pub const WINDOW_Y: f32 = 600.0;
 
```
```suggestion
const WINDOW_X: f32 = 1000.0;
const WINDOW_Y: f32 = 600.0;

```

lframosferreira committed:
mais fix

lframosferreira committed:
mais fixes

lframosferreira committed:
adicionei spawn do score

lframosferreira committed:
update score e similares so acontece se no estado in game

lframosferreira committed:
tentando fazer o chao se mover de maneira fluida junto com os obstaculos

lframosferreira committed:
mova chao, imediatamente

lframosferreira committed:
fmt

lframosferreira committed:
movimento do chao

igorlfs commented on dino/src/game/obstacle/systems.rs:
```diff
@@ -53,10 +60,12 @@ pub fn spawn_obstacles_over_time(
-    for (obstacle_transform, obstacle_entity) in obstacle_query.iter_mut() {
-        if obstacle_transform.translation.x < 0.0 {
+    for (obstacle_handle_image, obstacle_transform, obstacle_entity) in obstacle_query.iter_mut() {
+        let obstacle_width: f32 = assets.get(obstacle_handle_image).unwrap().size().x;
```
Esse unwrap pode causar um crash. Estava acontecendo comigo, mas não encontrei uma maneira confiável de reproduzir. De qualquer maneira, acho que é melhor evitar o `unwrap` e fazer um `if let Some(...) = `

igorlfs committed:
privacidade, imports e clippy

igorlfs committed:
fix: reseta vertical_movement no game_over

se você morrer pulando, quando você reiniciar, o pulo ainda vai estar
acontecendo (sem o reset)

igorlfs committed:
fix: reseta dino_down no game_over

se você morrer agachado, quando você reiniciar, você não vai conseguir
pular (porque a função handle_jump verifica is_down)

igorlfs committed:
remove mut

lframosferreira committed:
resolvendo questao de unwrap em None em alguns pontos do jogo
